File: challenge/leet/easy/q506.py
'''
 * 506. Relative Ranks
 * https://leetcode.com/problems/relative-ranks/description/
'''
import unittest
from heapq import heappush, heappop
from typing import List


class Solution:
    def findRelativeRanks(self, score: List[int]) -> List[str]:
        def medal(ith: int):
            if ith > 2:
                return str(ith+1)
            elif ith == 2:
                return 'Bronze Medal'
            elif ith == 1:
                return 'Silver Medal'
            else:
                return 'Gold Medal'

        hq = []
        for x, v in enumerate(score):
            heappush(hq, (-v, x))

        ith = 0
        while len(hq) > 0:
            head = heappop(hq)
            score[head[1]] = medal(ith)
            ith += 1

        return score


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()
        self.assertEqual(["Gold Medal","Silver Medal","Bronze Medal","4","5"], s.findRelativeRanks([5,4,3,2,1]))
        self.assertEqual(["18","Silver Medal","17","16","15","14","13","12","11","10","9","8","7","6","5","4","Bronze Medal","Gold Medal"],
                         s.findRelativeRanks([0,16,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,17]))


if __name__ == '__main__':
    unittest.main()

File: challenge/leet.gcc/easy/q506/q506.cpp
/**
 * 506. Relative Ranks
 * https://leetcode.com/problems/relative-ranks/description/ 
 */

#include <vector>
#include <string>
#include <iostream>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    string tostr(int i) {
        if (i == 0)
            return string("Gold Medal");
        else if (i == 1)
            return string("Silver Medal");
        else if (i == 2)
            return string("Bronze Medal");
        else
            return to_string(i+1);
    }

    vector<string> findRelativeRanks(vector<int>& score) {
        vector<pair<int, int>> h;
        for (int s = 0; s < score.size(); s++) {
            h.push_back(pair<int, int>{score[s], s});
            push_heap(h.begin(), h.end(), [](pair<int, int>& a, pair<int, int> &b){
                return a.first < b.first;
            });
        }

        vector<string> res(score.size());
        int i = 0;
        while (h.size() > 0) {
            res[h.front().second] = tostr(i);
            i++;
            pop_heap(h.begin(), h.end());
            h.pop_back();
        }
        return res;
    }
};

string reduce(vector<string> v) {
    return v.size() > 0 ? reduce(
        next(v.begin()), v.end(), string("[") + v[0], [](string a, string b) { return move(a) + ", " + b; })
        + string("]")
        : "[]";
}

int main() {
    Solution s;
    vector<int> scores;

    scores = vector<int>{ 5,4,3,2,1 };
    cout << "[Gold Medal, Silver Medal, Bronze Medal, 4, 5] ?" << endl
        << reduce(s.findRelativeRanks(scores)) << endl;

    scores = vector<int>{ 10,3,8,9,4 };
    cout << "[Gold Medal, 5, Bronze Medal, Silver Medal, 4] ?" << endl
        << reduce(s.findRelativeRanks(scores)) << endl;

    return 0;
}